demo/models/orderbytag.py

```diff
@@ -0,0 +1,214 @@
+            return [Invite]
+        if room.join_state == Quotient.JoinState.Leave:
+            return [Left]
+        
+        tags = self.get_filtered_tags(room)
+        if len(tags) == 0:
+            tags.insert(0, Untagged)
+        # Check successors, reusing room as the current frame, and for each group
+        # shadow this room if there's already any of its successors in the group
+        successor_room = room.successor() # TODO: Quotient.JoinState.Join
+        while successor_room is not None:
+            successor_tags = self.get_filtered_tags(successor_room)
+
+            if len(successor_tags) == 0:
+                tags.remove(Untagged)
+            else:
+                for tag in successor_tags:
+                    if tag in tags:
+                        tags.remove(tag)
+            
+            if len(tags) == 0:
+                return [] # No remaining groups, hide the room
+
+            successor_room = room.successor() # TODO: Quotient.JoinState.Join
+        return tags
+    
+    def get_filtered_tags(self, room: Quotient.Room) -> List[str]:
+        all_tags = room.tag_names
+        if room.is_direct_chat():
+            all_tags.append(DirectChat)
+        
+        result: List[str] = []
+        for tag in all_tags:
+            if find_index_with_wildcards(self.tags_order, '-' + tag) == len(self.tags_order):
+                result.append(tag) # Only copy tags that are not disabled
+        return result
+
+    def room_less_then(self, group_key: str, room1: Quotient.Room, room2: Quotient.Room):
+        if room1 == room2:
+            return False # 0. Short-circuit for coinciding room objects
+
+        # 1. Compare tag order values
+        tag = group_key
+        order1 = room1.tag(tag).order
+        order2 = room2.tag(tag).order
+        if type(order2) != type(order1):
+            return not order2 == None
+        
+        if order1 and order2:
+            # Compare floats; fallthrough if neither is smaller
+            if order1 < order2:
+                return True
+            
+            if order1 > order2:
+                return False
+
+        # 2. Neither tag order is less than the other; compare room display names
+        if room1.display_name != room2.display_name:
+            return room1.display_name < room2.display_name
+
+        # 3. Within the same display name, order by room id
+        # (typically the case when both display names are completely empty)
+        if room1.id != room2.id:
+            return room1.id < room2.id
+        
+        # 4. Room ids are equal; order by connections (=userids)
+        connection1 = room1.connection
+        connection2 = room2.connection
+        if connection1 != connection2:
+            if connection1.user_id != connection2.user_id:
+                return connection1.user_id < connection2.user_id
+            
+            # 3a. Two logins under the same userid: pervert, but technically correct
```

Comment:
```suggestion
            # 4a. Two logins under the same userid: pervert, but technically correct
```
Also fixed now in the Quaternion source :)